x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -1043,7 +1052,13 @@ public DatafeedConfig build() {
             if (MlStrings.isValidId(id) == false) {
                 throw ExceptionsHelper.badRequestException(getMessage(INVALID_ID, ID.getPreferredName(), id));
             }
-            if (indices == null || indices.isEmpty() || indices.contains("")) {
+            if (esqlQuery != null && queryProvider != null) {
```

Comment:
`esqlQuery != null` also needs to be banned in the case where any node in the cluster is older than the version containing this change. You can call `ClusterState.getMinTransportVersion()` to find out, but then you'll need to have access to a `ClusterState` object.